trinity/plugins/builtin/ethstats/plugin.py

```diff
@@ -0,0 +1,48 @@
+import asyncio
+
+from argparse import (
+    ArgumentParser,
+    _SubParsersAction,
+)
+
+from trinity.extensibility import (
+    BaseEvent,
+    BasePlugin,
+    PluginContext,
+)
+from trinity.extensibility.events import (
+    TrinityStartupEvent,
+)
+
+from trinity.plugins.builtin.ethstats.stats_client import (
+    StatsClient,
+)
+
+
+class EthstatsPlugin(BasePlugin):
+
+    def __init__(self) -> None:
+        self.is_enabled: bool = False
+        # TODO: make server/secret configurable
+        self.stats_client: StatsClient = StatsClient('ws://localhost:3000/api', 'SECRET')
+
+    @property
+    def name(self) -> str:
+        return "Ethstats"
+
+    def configure_parser(self, arg_parser: ArgumentParser, subparser: _SubParsersAction) -> None:
+        arg_parser.add_argument(
+            "--ethstats",
+            action="store_true",
+            help="Enable node status reporting service (experimental)",
+        )
+
+    def handle_event(self, activation_event: BaseEvent) -> None:
+        if isinstance(activation_event, TrinityStartupEvent):
+            self.is_enabled = activation_event.args.ethstats
+
+    def should_start(self) -> bool:
+        return self.is_enabled
+
+    def start(self, context: PluginContext) -> None:
+        asyncio.ensure_future(self.stats_client.run())
```

Comment:
There's nothing yet. It's probably worth opening up an issue to collect specific use cases that need to be addressed. In this case, are you referring to how we would stop/restart this plugin or how it would recover from failure to auto-restart or something completely different?

Reply:
I'm interested in the clean-shutdown aspect (which might be what he meant too). In this case, there is no cancel token being passed to the stats client from any parent service, so nothing causes it to shut down smoothly.

I guess the plugin could listen for the shutdown event and call `await self.stats_client.cancel()`. Any other good options?

Reply:
Reimplemented whole plugin as BaseIsolatedPlugin, hopefully this should resolve issues with graceful shutdown.